File: Question/2_Add_Two_Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* cur = &dummy;

        int flag = 0; //进位标志
        while(l1 || l2) {
            int num = l1 ? l1->val : 0;
            num += l2 ? l2->val : 0;
            num += flag;
            flag = num / 10;

            cur->next = new ListNode(num % 10);
            cur = cur->next;
            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
        }
        if(flag) cur->next = new ListNode(flag);
        return dummy.next;
    }
};

File: Question_Go/2_Add_Two_Numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy := &ListNode{}
    cur := dummy
    carry := 0

    for l1 != nil || l2 != nil || carry != 0  {
        temp := 0
        if l1 != nil {
            temp += l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            temp += l2.Val
            l2 = l2.Next
        }
        temp += carry
        cur.Next = &ListNode{}
        cur = cur.Next
        cur.Val = temp % 10
        carry = temp / 10
    }

    return dummy.Next
}